🪨 feat: Add Bedrock Prompt Caching Support

dustinhealy opened the pull request:
## Summary

Closes #7109

Adds a new parameter, **Use Prompt Caching**, for Bedrock endpoints, enabling cache points to be automatically embedded in messages sent to [supported models](https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html#prompt-caching-models)

> Note: actual insertion logic relies on https://github.com/danny-avila/agents/pull/28 (as well as a bug-fix around tool calling in https://github.com/danny-avila/agents/pull/30). This PR now mainly concerns the addition of the **Use Prompt Caching** toggle itself as a parameter. Previous implementation details which have been moved and refactored for the agents repo (or are no longer necessary as a result of the refactor) are left with ~~strikethroughs~~ for the sake of historical tracking.

## Change Type

- [x] New feature (non-breaking change which adds functionality)

## Details

- ~~Introduced `addBedrockCacheControl` to insert cache points when **Use Prompt Caching** is enabled (`api/app/clients/prompts/addBedrockCacheControl.js`).~~
- ~~Integrated logic into the Bedrock agent client so cache control is applied as appropriate (`api/server/controllers/agents/client.js`).~~
- ~~Updated exports to make `addBedrockCacheControl` available (`api/app/clients/prompts/index.js`).~~
- Added `promptCache` (shown in UI as **Use Prompt Caching**) to Bedrock parameter settings and schema for configuration and validation (`packages/data-provider/src/parameterSettings.ts`, `packages/data-provider/src/bedrock.ts`).

## Testing

- Enabled and disabled **Use Prompt Caching** via parameter panel and verified the presence or absence of cache points in messages sent to Bedrock endpoints.
- Using LangFuse, verified traces contained expected cache read and cache write usage metadata when prompt caching turned on for an agent using the Bedrock endpoint.

## Test Configuration

Agent using AWS Bedrock endpoint with model: us.anthropic.claude-sonnet-4-20250514-v1:0 with **Use Prompt Caching** toggled on

### Three message chain with high-token initial message (1000+ token) in order to trigger cache reads and cache writes.

#### Initial Input
> **User**:
> Use the Calculator tool, what is 5123043/39201? This is a long input test to see if prompt caching works. Lorem ipsum Lorem ipsum Lorem ipsum... <Continues for around ~1300 more tokens>
>
> **AWS Bedrock Prompt Caching (Tool Use) Test Agent**:
> I'll calculate 5123043/39201 for you using the calculator tool.
> \<Ran calculator\>
> The result of 5123043/39201 is approximately 130.687.

#### Usage Metadata:
```JSON
{
    "inputTokens": 1794,
    "outputTokens": 26,
    "serverToolUsage": {},
    "totalTokens": 1820
}
```

---

#### Follow-up (triggers write to cache)
> **User**: 
> This is a follow-up to check for cache_write
>
> **AWS Bedrock Prompt Caching (Tool Use) Test Agent**:
> I understand you're testing the cache functionality with this follow-up message. The previous calculation of 5123043/39201 gave us the result of approximately 130.687.
> 
> Is there anything else you'd like me to calculate or help you with?

#### Usage Metadata:
```JSON
{
    "cacheReadInputTokenCount": 0,
    "cacheReadInputTokens": 0,
    "cacheWriteInputTokenCount": 1844,
    "cacheWriteInputTokens": 1844,
    "inputTokens": 4,
    "outputTokens": 58,
    "serverToolUsage": {},
    "totalTokens": 1906
}
```

---

#### Final follow-up (triggers read from cache)
> **User**:
> This is a follow-up to check for cache_read
>
> **AWS Bedrock Prompt Caching (Tool Use) Test Agent**:
> I see you're continuing to test the cache functionality. The calculation we performed earlier (5123043/39201 ≈ 130.687) should be available from the cached context along with all the Lorem ipsum text from your original message.
>
> Is there anything specific you'd like me to help you with or any other calculations you need?

#### Usage Metadata:
```JSON
{
    "cacheReadInputTokenCount": 1735,
    "cacheReadInputTokens": 1735,
    "cacheWriteInputTokenCount": 182,
    "cacheWriteInputTokens": 182,
    "inputTokens": 4,
    "outputTokens": 75,
    "serverToolUsage": {},
    "totalTokens": 1996
}
```

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] A pull request for updating the documentation has been submitted.

KiGamji commented:
@dustinhealy Hi there! Do you have any due dates for this? Thanks in advance

dustinhealy commented:
> @dustinhealy Hi there! Do you have any due dates for this? Thanks in advance

Hi KiGamji, still need to write tests but planning to get this ready for review within the next couple days.

KiGamji commented:
Hopefully we'll see this merged in the next week 👍

danny-avila commented:
> Hopefully we'll see this merged in the next week 👍

It requires an upstream change for it to fully work: 

https://github.com/langchain-ai/langchainjs/pull/8524

So we have to wait when that PR is tested/merged

mikhailmokhov commented:
> > Hopefully we'll see this merged in the next week 👍
> 
> It requires an upstream change for it to fully work:
> 
> [langchain-ai/langchainjs#8524](https://github.com/langchain-ai/langchainjs/pull/8524)
> 
> So we have to wait when that PR is tested/merged

Note that the [mentioned PR](https://github.com/langchain-ai/langchainjs/pull/8524) is adding cache point support only to `@langchain/aws` package, make sure you're using it. Also, [cachePoint](https://github.com/danny-avila/LibreChat/pull/8271/files#diff-e4933682cbf1dd2d69de3174f75d36827cef856b5d95fc2f64d0d4ffe37d393bR24) must be a separate content block in the content array, see example: https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html#prompt-caching-get-started. You have it as a text block property.

danny-avila commented on api/app/clients/prompts/addBedrockCacheControl.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {Array<BaseMessage>} messages - The array of message objects.
+ * @returns {Array<BaseMessage>} - The updated array of message objects with cache points added.
+ */
+function addBedrockCacheControl(messages) {
```
Move to `packages/api/format`

danny-avila commented on api/app/clients/prompts/index.js:
```diff
@@ -1,4 +1,5 @@
-const addCacheControl = require('./addCacheControl');
+const { addCacheControl } = require('./addCacheControl');
+const { addBedrockCacheControl } = require('./addBedrockCacheControl');
```
remove from `/api` workspace

danny-avila commented:
Lastly, rebase branch to latest (to resolve conflicts)

dustinhealy committed:
feat: Add Bedrock Cache Control Functionality

- fix: Update Bedrock Cache Control to Require cachePoint as a Separate Content Block

- Modified the addBedrockCacheControl function to ensure cachePoint is added as a separate content block in the content array, rather than as a property of text objects.

- refactor: move addBedrockCacheControl over to packages/api

- ci: add tests for addBedrockCacheControl until full coverage reached

Pfaufisch commented:
`@langchain/aws` `0.1.14` has been released and includes prompt caching: https://github.com/langchain-ai/langchainjs/pull/8524#issuecomment-3197615451

KiGamji commented:
@danny-avila @dustinhealy tagging for visibility too. We would like to see this, leading to huge cost savings! Thanks in advance for your work!

dustinhealy committed:
ci: add test similar to example from the langchain PR

KiGamji commented:
any reason for this not being merged?

KiGamji commented:
yeah, we are waiting for this too, as much as new Bedrock keys format support, and fixes for the PROXY env!

AetherMagee commented:
Are there any blockers for this besides needing a rebase?

Defiect commented:
> > Are there any blockers for this besides needing a rebase?
> 
> The agents repo still needs a version bump for @langchain/aws in order for this feature to actually work + we still have to do some testing on the Bedrock side of things to make sure the prompt caching points are working as intended. I will rebase later today after I finish my current PR just to resolve the merge conflicts. Thanks all for your interest, we know this is highly anticipated feature, so we appreciate your patience in the meantime!

Awesome, thank you. Your  effort is greatly appreciated!

matthill33 commented:
> > > Are there any blockers for this besides needing a rebase?
> > 
> > 
> > The agents repo still needs a version bump for @langchain/aws in order for this feature to actually work + we still have to do some testing on the Bedrock side of things to make sure the prompt caching points are working as intended. I will rebase later today after I finish my current PR just to resolve the merge conflicts. Thanks all for your interest, we know this is highly anticipated feature, so we appreciate your patience in the meantime!
> 
> Awesome, thank you. Your effort is greatly appreciated!

I'm also very much looking forward to this feature being implemented, thank you @dustinhealy for your work on it!

dvejsada commented:
> @danny-avila Any chance to get this merged until October? I appreciate your time and effort in maintaining this huge project, but this PR is already two months old and it would need more effort for @dustinhealy to rebase and maintain it each time, unless it gets merged. Also, this is one of the mostly anticipated features by activity, along with conversation cost tracking (#9190, #8531), conversation pinning (#7550), PassKeys & E2EE, so I really hope you will give it some attention in the neartime. Thanks in advance!

Very anticipated indeed!;)

ldavis9000aws commented:
Ok...I'm sure there are more important things the project is focused on, but just a reminder:

<img width="1420" height="1537" alt="image" src="https://github.com/user-attachments/assets/372deb6c-f8fe-4006-ae76-7b871db73d1c" />

KiGamji commented:
With all the respect to Danny, I absolutely agree with this :/
It has been 3 months...

KiGamji commented:
@danny-avila please, can you review this?

KiGamji commented:
It surely is, system prompts don't even work for Bedrock it seems... The amount of hyperfocus on agents is too big over the last months to be honest, and it's frustrating to see dozens of pull requests just being constantly rebased or unreviewed and completely ignored even over pings

matthill33 commented:
Would love for this to be reviewed and included in the next release

ldavis9000aws commented:
@danny-avila  In all seriousness Danny, please give us a status on this PR. I get that you're busy, have other priorities, etc., but there have been zero updates on this.

KiGamji commented:
this project should be renamed AgentChat at this point

<img width="2095" height="1434" alt="image" src="https://github.com/user-attachments/assets/def732fd-40f5-4e38-b9ff-9ac564048187" />

KiGamji commented:
seriously though, I don't get why would you keep it dangling like this, system prompts are broken, PROXY environment variable is broken, prompt caching is unmerged for 4 months, the AWS Bedrock is basically unmaintained. this is straight up just wasting the time of people who are new to this project, they should consider using the Anthropic endpoint natively then, or writing a convert service, which would do the caching job itself (which we did with docker compose to use Vertex API via Anthropic endpoint with Prompt Caching and other features leading to insane cost savings, especially in agentic usage purposes).

Ziyann commented:
> and I also recommend switching to Vertex API for other people here—it is many times faster and painless, we would personally never return back to Bedrock. (assuming that writing a simple router service to convert and pass requests is a not a big deal for some people here)

I have https://github.com/danny-avila/LibreChat/pull/9798 pending to avoid needing a router service for Vertex AI, waiting for Danny to give me feedback to help merging it.

I'm not really sure what can I say, I'm glad that the project is open source and maintained, but I'm also frustrated that agents are getting all the focus, at the cost of neglecting other, much simpler wishes/features, especially PRs that community members put work into. We're happy to contribute code, but I'm sure people are getting discouraged when they work on something, submit a PR and it becomes ignored.

KiGamji commented:
A huge number of people and their users are waiting while being fed false promises.

dustinhealy committed:
refactor: move addBedrockCacheControl logic and tests to agents repository

danny-avila committed:
chore: remove extraneous comment

danny-avila committed:
chore: update @librechat/agents dependency to version 3.0.12

danny-avila committed:
chore: update @librechat/agents dependency to version 3.0.13

danny-avila committed:
chore: update @librechat/agents dependency to version 3.0.14

danny-avila committed:
chore: update @librechat/agents to v3.0.15

danny-avila committed:
chore: update default value for prompt cache setting to true

danny-avila committed:
refactor: set default promptCache to true for claude and nova models
